File: examples/governance_static_go/domain/training_record.go
package domain

import (
	"time"
	"github.com/google/uuid"
)

type TrainingRecord struct {
	ID string `json:"id"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
	StakeholderId string `json:"stakeholder_id"`
	PolicyId string `json:"policy_id"`
	CompletedAt time.Time `json:"completed_at"`
	ExpiresAt time.Time `json:"expires_at"`
	Certification string `json:"certification"`
	Status string `json:"status"`
}

func NewTrainingRecord(stakeholderId string, policyId string, completedAt time.Time, expiresAt time.Time, certification string, status string) *TrainingRecord {
	a := &TrainingRecord{
		ID:        uuid.New().String(),
		StakeholderId: stakeholderId,
		PolicyId: policyId,
		CompletedAt: completedAt,
		ExpiresAt: expiresAt,
		Certification: certification,
		Status: status,
		CreatedAt: time.Now(),
		UpdatedAt: time.Now(),
	}
	return a
}

func (a *TrainingRecord) Validate() error {
	if a.StakeholderId == "" {
		return &ValidationError{Field: "stakeholder_id", Message: "stakeholder_id can't be blank"}
	}
	if a.PolicyId == "" {
		return &ValidationError{Field: "policy_id", Message: "policy_id can't be blank"}
	}
	// invariant: expires_at must be after completed_at
	return nil
}

File: examples/governance/compliance_domain/lib/compliance_domain/training_record/training_record.rb
require 'hecks/mixins/model'

module ComplianceDomain
  class TrainingRecord
    autoload :Lifecycle, "compliance_domain/training_record/lifecycle"

    include Hecks::Model

    attribute :stakeholder_id
    attribute :policy_id
    attribute :completed_at
    attribute :expires_at
    attribute :certification
    attribute :status

    # State predicates — see lifecycle.rb for full state machine
    def assigned?; status == "assigned"; end
    def completed?; status == "completed"; end

    private

    def validate!
      raise ValidationError.new("stakeholder_id can't be blank", field: :stakeholder_id, rule: :presence) if stakeholder_id.nil? || (stakeholder_id.respond_to?(:empty?) && stakeholder_id.empty?)
      raise ValidationError.new("policy_id can't be blank", field: :policy_id, rule: :presence) if policy_id.nil? || (policy_id.respond_to?(:empty?) && policy_id.empty?)
    end

    def check_invariants!
      raise InvariantError, "expires_at must be after completed_at" unless instance_eval(&proc { !expires_at || !completed_at || expires_at.to_s >= completed_at.to_s[0, 10] })
    end
  end
end
